[AArch64] Improve post-inc stores of SIMD/FP values

guy-david opened the pull request:
Add patterns to match post-increment truncating stores from lane 0 of wide integer vectors (v4i32/v2i64) to narrower types (i8/i16/i32). This avoids transferring the value through a GPR when storing.

Also remove the pre-legaliztion early-exit in `combineStoreValueFPToInt` as it prevented the optimization from applying in some cases.

paulwalker-arm commented:
You cannot fix it like this because the patterns are not really missing.  It is a design decision for element types smaller than i32 to use i32 as their container type.  This is common across element extracts, inserts, splats and likely elsewhere.

It might be worth going the isel route you tried before but either restricting the pattern to the zero index case or implement a way to ensure the index is scaled appropriately for the target instruction.

guy-david committed:
[AArch64] Improve post-inc stores of SIMD/FP values

Add patterns to match post-increment truncating stores from lane 0 of
wide integer vectors (v4i32/v2i64) to narrower types (i8/i16/i32).
This avoids transferring the value through a GPR when storing.

Also remove the pre-legalization early-exit in combineStoreValueFPToInt
as it prevented the optimization from applying in some cases.

paulwalker-arm reviewed (approved):
FYI: It's generally preferred to add the new tests separately as the first commit of the PR. That way it makes it easier to see the effect of the code changes in the rest of the PR.

guy-david commented:
Re-opening, as it seems there's a GitHub bug due to the automatic deletion of the original branch.
